packages/ts/react-grid/src/autogrid.tsx

```diff
@@ -78,27 +84,83 @@ function createDataProvider<TItem>(
   };
 }
 
-function createColumns(model: ModelConstructor<unknown, AbstractModel<unknown>>, visibleColumns?: string[]) {
+function createColumns(
+  model: ModelConstructor<unknown, AbstractModel<unknown>>,
+  setPropertyFilter: (propertyFilter: PropertyStringFilter) => void,
+  options: { visibleColumns?: string[]; headerFilters?: boolean },
+) {
   const properties = getProperties(model);
-  const effectiveColumns = visibleColumns ?? properties.map((p) => p.name);
+  const effectiveColumns = options.visibleColumns ?? properties.map((p) => p.name);
   const effectiveProperties = effectiveColumns
     .map((name) => properties.find((prop) => prop.name === name))
     .filter(Boolean) as PropertyInfo[];
 
-  return effectiveProperties.map((p) => (
-    <GridSortColumn path={p.name} header={p.humanReadableName} key={p.name} autoWidth></GridSortColumn>
-  ));
+  return effectiveProperties.map((p) => {
+    const column = <GridSortColumn path={p.name} header={p.humanReadableName} key={p.name} autoWidth></GridSortColumn>;
+    if (options.headerFilters) {
+      const headerRenderer = useCallback(
```

Comment:
Using a hook like `useCallback` here is problematic, as this can easily lead to a different number of hook calls during rendering, which is not allowed. This can happen for example when changing the visible columns or turning header filters on or off. 

Alternatively we could:
- Create a single header renderer for all columns and memoize that once outside of any loops or conditions. The renderer would then need to access column-specific data (like property name) from the column instance itself which should already be passed in the props of the header renderer somehow. To store the property name we could use a data attribute.
- Or keep creating a separate renderer for each column. Then they would need to be stored in a map, which in turn is stored in a `useMemo` or a ref somewhere outside of any loops or conditions.

Another issue of memoizing the renderers is that the `setPropertyFilter` reference is always the initial one, it will never be updated on subsequent renders. That also means that `setPropertyFilter` will always reference the initial `internalFilter` state. That currently works, because the function mutates the filter object. But if we would add some other feature that calls `setInternalFilter`, then calling `setPropertyFilter` from one of the header renderers would undo that change.

Reply:
Changed to use one renderer. Does that resolve the problems?

Reply:
The header renderer needs to be memoized in some way because of https://github.com/vaadin/react-components/issues/146

Reply:
Is there a standard react way for handling communication with the singleton renderer then? Should the headerFilterRenderer be stored as state?

Reply:
The current hook usage looks safe due to the change to `useColumns` where you're signaling that it must be called everytime, and if someone want's to change that they will have to deal with refactoring it's content's too.

`setPropertyFilter` is still an issue, as the memoized header renderer still only get's the reference to the initial version of that function, which aways references the initial `internalFilter` state. To fix that you'd have to put `setPropertyFilter` into a ref on every render so that the header renderer can access the latest reference from there. The safest way would probably be to move all inputs of the renderer into a ref - for example when switching models, the renderer would need to access the updated `properties`, which it could only through a ref.

Reply:
Change property filter to be a ref and also the properties and added test that the model can be changed, the header filters dynamically turned on and off, and that filters also work after a model change